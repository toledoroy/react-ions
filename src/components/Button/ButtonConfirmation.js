import React, { Component} from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames/bind'
import optclass from '../internal/OptClass'
import Button from './Button'
import style from './style.scss'

export class ButtonConfirmation extends Component {
  constructor(props) {
    super(props)
    this.mql = window.matchMedia('(max-width: 992px)')
  }

  static defaultProps = {
    prompt: 'Are you sure?'
  }

  static propTypes = {
    /**
     * Optional styles to add to the button.
     */
    optClass: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.string
    ]),
    /**
     * The size of button.
     */
    size: PropTypes.string,
    /**
     * Whether the button is disabled.
     */
    disabled: PropTypes.bool,
    /**
     * Whether to display only an icon on small screens
     */
    collapse: PropTypes.bool,
    /**
     * Position of the button on the page for overlay caret orientation when the screen size is tablet-sized or smaller
     */
    position: PropTypes.oneOf(['left', 'right']),
    /**
     * Text that will appear in the overlay prompt.
     */
    prompt: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /**
     * Function used to pass up the confirmation to the parent component
     */
    handleConfirmation: PropTypes.func
  }

  state = {
    confirmationOverlayOpen: false,
    confirmationOverlayOffset: 0,
    hasBeenOpened: false,
    wide: false
  }

  handleOpen = () => {
    if (!this.props.hasBeenOpened) {
      this.setState({
        hasBeenOpened: true
      }, () => {
        this.handleSetup()
      })
    }
    this.setState({
      confirmationOverlayOpen: !this.state.confirmationOverlayOpen
    })
  }

  handleConfirmation = confirm => {
    this.setState({ confirmationOverlayOpen: false }, () => {
      if (typeof this.props.handleConfirmation === 'function' && confirm) {
        this.props.handleConfirmation(confirm)
      }
    })
  }

  handleSetup = () => {
    const triggerRect = this._trigger.children[0].getBoundingClientRect()
    const overlayRect = this._trigger.querySelector("[class*='confirmation-overlay']").getBoundingClientRect()

    this.setState({
      triggerWidth: triggerRect.width,
      overlayWidth: overlayRect.width
    })
  }

  handleWide = () => {
    if (this.props.prompt.length > 25) {
      this.setState({ wide: true })
    }
  }

  getStyles = () => {
    if (this.props.position === 'right') {
      return {
        right: `${this.state.confirmationOverlayOffset}px`
      }
    }
    if (this.props.position === 'left') {
      return {
        left: `${this.state.confirmationOverlayOffset}px`
      }
    }

    // https://developers.google.com/web/updates/2016/06/absolute-positioned-children
    // Once ^ is supported in Safari and Firefox we can remove this and allow flex box to do its thing
    return {
      left: `-${((this.state.overlayWidth - this.state.triggerWidth) / 2)}px`
    }
  }

  getCaretStyles = () => {
    if (this.props.position === 'right') {
      return { right: `calc(${(this.state.triggerWidth / 2) + 7.5}px)`}
    }
    if (this.props.position === 'left') {
      return { left: `calc(${(this.state.triggerWidth / 2) - 5}px)` }
    }
    return this.state.wide ? { left: '95px' } : { left: '75px' }
  }

  /**
   * To trigger JavaScript changes on a breakpoint
   * @param  {Object} mediaQueryList
   * @return {Boolean} whether the viewport is < or > than max-width 768px
   */
  handleMediaChange = () => {
    this.handleSetup()
  }

  componentDidMount = () => {
    this.mql.addListener(this.handleMediaChange)
    this.handleSetup()
    this.handleWide()
  }

  componentWillUnmount = () => {
    this.mql.removeListener(this.handleMediaChange)
  }

  shouldComponentUpdate = (nextProps, nextState) => {
    if (nextProps.position !== this.props.position) return true
    if (nextState.wide !== this.state.wide) return true
    if (nextState.triggerWidth !== this.state.triggerWidth) return true
    if (nextState.overlayWidth !== this.state.overlayWidth) return true
    if (nextState.confirmationOverlayOpen !== this.state.confirmationOverlayOpen) return true
    if (nextProps.prompt !== this.props.prompt) return true
    if (nextProps.loading !== this.props.loading) return true

    return false
  }

  render = () => {
    const cx = classNames.bind(style)
    const { ...other } = this.props

    const buttonClass = optclass(style, ['confirmation-approve-btn'])
    const overlayPositionClass = this.props.position ? style[this.props.position] : null
    const overlayVisibleClass = this.state.confirmationOverlayOpen ? style['visible'] : null
    const overlayWideClass = this.state.wide ? style['wide'] : null
    const confirmationOverlayClasses = cx(overlayVisibleClass, overlayPositionClass, overlayWideClass, style['confirmation-overlay'])

    return (
      <div ref={trigger => this._trigger = trigger} className={style['confirmation-wrapper']}>
        <Button {...other} disabled={this.state.confirmationOverlayOpen} onClick={this.handleOpen}>
          { this.props.children }
        </Button>
        <div className={confirmationOverlayClasses} style={this.getStyles()}>
          <em style={this.getCaretStyles()}></em>
          <span className={style['confirmation-text']}>{this.props.prompt}</span>
          <div className={style['button-wrapper']}>
            <Button onClick={this.handleConfirmation.bind(this, false)} optClass='danger-alt'>Cancel</Button>
            <Button onClick={this.handleConfirmation.bind(this, true)} optClass={buttonClass}>Yes</Button>
          </div>
        </div>
      </div>
    )
  }
}

export default ButtonConfirmation
