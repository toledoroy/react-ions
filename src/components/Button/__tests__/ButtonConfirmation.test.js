import React from 'react'
import ButtonConfirmation from '../ButtonConfirmation'

describe('ButtonConfirmation', () => {
  let wrapper, inst

  beforeEach(() => {
    window.matchMedia = () => ({
      addListener: sinon.spy(),
      removeListener: sinon.spy()
    })
  })

  afterEach(() => {
    window.matchMedia = null
  })

  describe('render', () => {
    it('displays confirmation button without prompt', () => {
      expect(mount(<ButtonConfirmation>Delete</ButtonConfirmation>)).toMatchSnapshot()
    })

    it('handles collapse with confirmation button', () => {
      expect(mount(<ButtonConfirmation collapse={true}>Delete</ButtonConfirmation>)).toMatchSnapshot()
    })
  })

  describe('handleOpen', () => {
    it('handles opening and closing on the confirmation button', () => {
      wrapper = mount(<ButtonConfirmation hasBeenOpened={true}>Delete</ButtonConfirmation>)
      inst = wrapper.instance()

      inst.handleOpen()
      expect(wrapper.state().confirmationOverlayOpen).toBeTruthy()

      inst.handleOpen()
      expect(wrapper.state().confirmationOverlayOpen).toBeFalsy()
    })
  })

  describe('handleConfirmation', () => {
    it('handles confirmation on the confirmation button', () => {
      const handleConfirmation = sinon.spy()
      wrapper = mount(<ButtonConfirmation handleConfirmation={handleConfirmation}>Delete</ButtonConfirmation>)
      wrapper.setState({ confirmationOverlayOpen: true })

      wrapper.instance().handleConfirmation()
      expect(wrapper.state().confirmationOverlayOpen).toBeFalsy()
      expect(handleConfirmation.called).toBeFalsy()
    })

    it('calls the handleConfirmation prop if it is a function and if true is passed to the method', () => {
      const handleConfirmation = sinon.spy()
      wrapper = mount(<ButtonConfirmation handleConfirmation={handleConfirmation}>Delete</ButtonConfirmation>)
      wrapper.setState({ confirmationOverlayOpen: true })

      wrapper.instance().handleConfirmation(true)
      expect(wrapper.state().confirmationOverlayOpen).toBeFalsy()
      expect(handleConfirmation.called).toBeTruthy()
    })
  })

  describe('getStyles', () => {
    it('get styles on the confirmation button when position not passed', () => {
      wrapper = mount(<ButtonConfirmation>Delete</ButtonConfirmation>)
      wrapper.setState({
        overlayWidth: 20,
        triggerWidth: 10
      })

      expect(wrapper.instance().getStyles()).toMatchSnapshot()
    })

    it('get styles on the confirmation button when position right', () => {
      wrapper = mount(<ButtonConfirmation position='right'>Delete</ButtonConfirmation>)
      wrapper.setState({confirmationOverlayOffset: 40})

      expect(wrapper.instance().getStyles()).toMatchSnapshot()
    })

    it('get styles on the confirmation button when position left', () => {
      wrapper = mount(<ButtonConfirmation position='left'>Delete</ButtonConfirmation>)
      wrapper.setState({confirmationOverlayOffset: 40})

      expect(wrapper.instance().getStyles()).toMatchSnapshot()
    })
  })

  describe('getCaretStyles', () => {
    it('get caret styles on the confirmation button when position not passed and is not wide', () => {
      wrapper = mount(<ButtonConfirmation>Delete</ButtonConfirmation>)
      wrapper.setState({ wide: false })

      expect(wrapper.instance().getCaretStyles()).toMatchSnapshot()
    })

    it('get caret styles on the confirmation button when position not passed and is wide', () => {
      wrapper = mount(<ButtonConfirmation>Delete</ButtonConfirmation>)
      wrapper.setState({ wide: true })

      expect(wrapper.instance().getCaretStyles()).toMatchSnapshot()
    })

    it('get caret styles on the confirmation button when position right', () => {
      wrapper = mount(<ButtonConfirmation position={'right'}>Delete</ButtonConfirmation>)
      wrapper.setState({triggerWidth: 40})

      expect(wrapper.instance().getCaretStyles()).toMatchSnapshot()
    })

    it('get caret styles on the confirmation button when position left', () => {
      wrapper = mount(<ButtonConfirmation position={'left'}>Delete</ButtonConfirmation>)
      wrapper.setState({triggerWidth: 40})

      expect(wrapper.instance().getCaretStyles()).toMatchSnapshot()
    })
  })

  describe('handleSetup', () => {
    it('handles setup on the confirmation button', () => {
      wrapper = mount(<ButtonConfirmation>Delete</ButtonConfirmation>)
      wrapper.instance()._trigger = {
        children: [{
          getBoundingClientRect: () => {
            return {
              width: 40
            }
          }
        }],
        querySelector: () => {
          return {
            getBoundingClientRect: () => {
              return {
                width: 40
              }
            }
          }
        }
      }

      wrapper.instance().handleSetup()
      expect(wrapper.state().triggerWidth).toEqual(40)
      expect(wrapper.state().overlayWidth).toEqual(40)
    })
  })

  describe('handleWide', () => {
    it('handles wide on the confirmation button', () => {
      wrapper = mount(<ButtonConfirmation prompt={'This is a test.'}>Delete</ButtonConfirmation>)
      inst = wrapper.instance()

      inst.handleWide()
      expect(wrapper.state().wide).toBeFalsy()

      wrapper.setProps({ prompt: 'This is a test that will make it wide' })
      wrapper.update()

      // Mocking handleSetup because it was already tested and requires a decent setup
      inst.handleSetup = sinon.spy()

      inst.handleWide()
      expect(wrapper.state().wide).toBeTruthy()
    })
  })

  describe('shouldComponentUpdate', () => {
    let state, shouldComponentUpdate
    const props = {
      size: 'large',
      disabled: false,
      loading: false,
      collapse: false,
      optClass: 'test-class',
      onClick: () => {},
      prompt: 'Test prompt'
    }

    beforeEach(() => {
      wrapper = mount(<ButtonConfirmation {...props}>Delete</ButtonConfirmation>)
      state = wrapper.state()
      shouldComponentUpdate = wrapper.instance().shouldComponentUpdate
    })

    it('does not update when props and state have not changed', () => {
      expect(shouldComponentUpdate(props, state)).toBeFalsy()
    })

    it('updates when loading is changed', () => {
      expect(shouldComponentUpdate({ ...props, loading: true }, state)).toBeTruthy()
    })

    it('updates when prompt is changed', () => {
      expect(shouldComponentUpdate({ ...props, prompt: 'New prompt' }, state)).toBeTruthy()
    })

    it('updates when confirmationOverlayOpen is changed', () => {
      expect(shouldComponentUpdate(props, { ...state, confirmationOverlayOpen: true })).toBeTruthy()
    })

    it('updates when triggerWidth is changed', () => {
      expect(shouldComponentUpdate(props, { ...state, triggerWidth: 10 })).toBeTruthy()
    })

    it('updates when overlayWidth is changed', () => {
      expect(shouldComponentUpdate(props, { ...state, overlayWidth: 50 })).toBeTruthy()
    })

    it('updates when wide is changed', () => {
      expect(shouldComponentUpdate(props, { ...state, wide: true })).toBeTruthy()
    })
  })
})
