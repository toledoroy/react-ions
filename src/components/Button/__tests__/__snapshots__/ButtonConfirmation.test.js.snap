// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonConfirmation getCaretStyles get caret styles on the confirmation button when position left 1`] = `
Object {
  "left": "calc(15px)",
}
`;

exports[`ButtonConfirmation getCaretStyles get caret styles on the confirmation button when position not passed and is not wide 1`] = `
Object {
  "left": "75px",
}
`;

exports[`ButtonConfirmation getCaretStyles get caret styles on the confirmation button when position not passed and is wide 1`] = `
Object {
  "left": "95px",
}
`;

exports[`ButtonConfirmation getCaretStyles get caret styles on the confirmation button when position right 1`] = `
Object {
  "right": "calc(27.5px)",
}
`;

exports[`ButtonConfirmation getStyles get styles on the confirmation button when position left 1`] = `
Object {
  "left": "40px",
}
`;

exports[`ButtonConfirmation getStyles get styles on the confirmation button when position not passed 1`] = `
Object {
  "left": "-5px",
}
`;

exports[`ButtonConfirmation getStyles get styles on the confirmation button when position right 1`] = `
Object {
  "right": "40px",
}
`;

exports[`ButtonConfirmation render displays confirmation button without prompt 1`] = `
<ButtonConfirmation
  prompt="Are you sure?"
>
  <div
    className="confirmation-wrapper"
  >
    <Button
      disabled={false}
      onClick={[Function]}
      prompt="Are you sure?"
      type="button"
    >
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        <em>
          Delete
        </em>
      </button>
    </Button>
    <div
      className="confirmation-overlay"
      style={
        Object {
          "left": "-0px",
        }
      }
    >
      <em
        style={
          Object {
            "left": "75px",
          }
        }
      />
      <span
        className="confirmation-text"
      >
        Are you sure?
      </span>
      <div
        className="button-wrapper"
      >
        <Button
          onClick={[Function]}
          optClass="danger-alt"
          type="button"
        >
          <button
            className="btn danger-alt"
            onClick={[Function]}
            type="button"
          >
            <em>
              Cancel
            </em>
          </button>
        </Button>
        <Button
          onClick={[Function]}
          optClass="confirmation-approve-btn"
          type="button"
        >
          <button
            className="btn confirmation-approve-btn"
            onClick={[Function]}
            type="button"
          >
            <em>
              Yes
            </em>
          </button>
        </Button>
      </div>
    </div>
  </div>
</ButtonConfirmation>
`;

exports[`ButtonConfirmation render handles collapse with confirmation button 1`] = `
<ButtonConfirmation
  collapse={true}
  prompt="Are you sure?"
>
  <div
    className="confirmation-wrapper"
  >
    <Button
      collapse={true}
      disabled={false}
      onClick={[Function]}
      prompt="Are you sure?"
      type="button"
    >
      <button
        className="btn collapse"
        onClick={[Function]}
        type="button"
      >
        <em>
          Delete
        </em>
      </button>
    </Button>
    <div
      className="confirmation-overlay"
      style={
        Object {
          "left": "-0px",
        }
      }
    >
      <em
        style={
          Object {
            "left": "75px",
          }
        }
      />
      <span
        className="confirmation-text"
      >
        Are you sure?
      </span>
      <div
        className="button-wrapper"
      >
        <Button
          onClick={[Function]}
          optClass="danger-alt"
          type="button"
        >
          <button
            className="btn danger-alt"
            onClick={[Function]}
            type="button"
          >
            <em>
              Cancel
            </em>
          </button>
        </Button>
        <Button
          onClick={[Function]}
          optClass="confirmation-approve-btn"
          type="button"
        >
          <button
            className="btn confirmation-approve-btn"
            onClick={[Function]}
            type="button"
          >
            <em>
              Yes
            </em>
          </button>
        </Button>
      </div>
    </div>
  </div>
</ButtonConfirmation>
`;
